bgls/utils.py

```diff
@@ -56,3 +61,157 @@ def cirq_density_matrix_bitstring_probability(
         density_matrix_state.target_tensor, num_qubits
     )
     return np.abs(density_matrix[index, index])
+
+
+def cirq_stabilizer_ch_bitstring_probability(
+    stabilizer_ch_form_state: cirq.sim.StabilizerChFormSimulationState,
+    bitstring: str,
+) -> float:
+    """Returns the probability of measuring the `bitstring` (|z⟩) in the
+    `cirq.StabilizerChFormSimulationState` (U_C U_H|s⟩), i.e. |⟨z|ψ⟩|^2.
+
+    Args:
+        stabilizer_ch_form_state: Stabilizer state in CH form (arxiv:
+        1808.00128) as a 'cirq.StabilizerChFormSimulationState'.
+        bitstring: Bitstring |z⟩ as a binary string.
+
+    """
+    # the state is of type StabilizerStateChForm
+    state = stabilizer_ch_form_state.state
+    index = int(bitstring, 2)
+    # this runs in O(n^2) for an n qubit state
+    return np.abs(state.inner_product_of_state_and_x(index)) ** 2
+
+
+def apply_near_clifford_gate(
+    op: cirq.Operation,
+    state: bgls.simulator.State,
+    rng: np.random.RandomState = np.random.RandomState(),
+) -> None:
+    """
+    Applies a Clifford+T gate to a state. If the gate is Clifford,
+    apply normally, else choose one of the gates in the Clifford-expansion
+    to apply.
+
+    Parameters:
+        op: operation to apply.
+        state: simulator state to update.
+        rng: rng for controlling gate-expansion decisions.
+    """
+    if cirq.has_stabilizer_effect(op):
+        cirq.protocols.act_on(op, state)
+    else:
+        # assuming T gate
+        # assert isinstance(op,
+        #                   cirq.ops.common_gates.ZPowGate) and \
+        #        op.gate.exponent == 0.25
+        probs = np.power(
+            np.abs(
+                [
+                    np.cos(np.pi / 8) - np.sin(np.pi / 8),
+                    np.sqrt(2.0)
+                    * np.exp(-(0 + 1j) * np.pi / 4)
+                    * np.sin(np.pi / 8),
+                ]
+            ),
+            2,
+        )
+        cand_gates = [cirq.I(op.qubits[0]), cirq.S(op.qubits[0])]
+        chosen_gate = rng.choice(cand_gates, p=probs / sum(probs))
+        cirq.protocols.act_on(chosen_gate, state)
+
+
+def improved_random_circuit(
```

Comment:
Odd name. Use `get_random_circuit`, `generate_random_circuit`, etc.

Reply:
Updated this, sounds better now